### PR TITLE
feat: Implement pause() and resume()

### PR DESCRIPTION
**Pull request by lynnagara**
Add pause and resume methods for the Kafka consumer

### CODE REVIEW

**Comment by volokluev** [comment]
Is this code tested at all?

**Comment by lynnagara** [comment]
> Is this code tested at all?

No tests yet. The reason is that we need to finish implementing some other parts of this consumer for this to actually work. Specifically the part that tracks partitions and updates self.offsets in the assignment/revocation callback. Right now it's always empty so these functions will immediately error with ConsumerError::UnassignedPartition.

**Comment by volokluev** [comment]
A few questions:

Is implementing these necessary (or useful) for what we're trying to learn? What does it unlock for us in terms of gauging performance or ease of writing consumers? 

Assuming it is useful:
Should we implement the offset tracking first or is this a chicken/egg situation?

Do we need to implement offset tracking and rebalancing mechanics right now?

I understand that our consumers need to handle rebalancing but for the test we have in  mind, we'll probably only run one partition so I don't have a clear picture of how it helps us.

**Comment by lynnagara** [comment]
It's not just for rebalancing. The assignment callback is called when the first (and only partition) gets assigned (or revoked).

File: src/backends/mod.rs [hunk 1]
```diff
@@ -13,6 +13,9 @@ pub enum ConsumerError {
     #[error("End of partition reached")]
     EndOfPartition,
 
+    #[error("Invalid consumer state")]
+    InvalidConsumerState,
```
Comment (fpacifici) [inline src/backends/mod.rs:17]: I would add some details on what the invalid state was. 
You can make the error message parametric and provide some details 
https://docs.rs/thiserror/latest/thiserror/#example

File: src/backends/kafka/mod.rs [hunk 2]
```diff
@@ -205,17 +256,13 @@ impl<'a> ArroyoConsumer<'a, KafkaPayload> for KafkaConsumer {
...
-            KafkaConsumerState::NotSubscribed,
-        ]
-        .contains(&self.state)
-        {
-            return Err(ConsumerError::ConsumerClosed);
+        match self.state {
+            KafkaConsumerState::Closed => Err(ConsumerError::ConsumerClosed),
+            KafkaConsumerState::NotSubscribed | KafkaConsumerState::Error => {
+                Err(ConsumerError::InvalidConsumerState)
+            }
```
Comment (fpacifici) [inline src/backends/kafka/mod.rs:263]: I think ConsumerClosed is a special case of InvalidConsumerState. Why not having `InvalidConsumerState` and passing the state as a parameter for the error when we raise it ?

Comment (lynnagara) [reply src/backends/kafka/mod.rs]: Done. I do worry it is excessively verbose to write `Err(ConsumerError::InvalidConsumerState(InvalidState::Closed))` and I do think it might be better to just add all 3 kinds of errors (not subscribed, error and closed) to the list instead. What do you think/

Comment (lynnagara) [reply src/backends/kafka/mod.rs]: Switched to the 3 errors option. I think it's more straightforward this way.

File: src/backends/kafka/mod.rs [hunk 3]
```diff
@@ -178,30 +189,56 @@ impl<'a> ArroyoConsumer<'a, KafkaPayload> for KafkaConsumer {
...
     }
 
-    fn pause(&mut self, _: HashSet<Partition>) -> Result<(), ConsumerError> {
-        //TODO: Implement this
+    fn pause(&mut self, partitions: HashSet<Partition>) -> Result<(), ConsumerError> {
+        self.state.assert_consuming_state()?;
+
+        let mut topic_map = HashMap::new();
+        for partition in partitions {
+            let offset = *self
```
Comment (fpacifici) [inline src/backends/kafka/mod.rs:211]: Why did you need to explicitly dereference self here ?

**Review by fpacifici — APPROVED** [review]

File: src/backends/kafka/mod.rs (hunk 3 above)
Comment (lynnagara) [reply src/backends/kafka/mod.rs:211]: HashMap::get() on `self.offsets` returns an Option<&u64>. We copy the offset number into our new `topic_map`.
